src/linear_algebra/free_module.lean

```diff
@@ -0,0 +1,352 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+-/
+import linear_algebra.basis
+import linear_algebra.finsupp_vector_space
+import ring_theory.principal_ideal_domain
+
+/-! # Free modules
+
+A free `R`-module `M` is a module with a basis over `R`,
+equivalently it is an `R`-module linearly equivalent to `ι →₀ R` for some `ι`.
+
+This file proves a submodule of a free `R`-module of finite rank is also
+a free `R`-module of finite rank, if `R` is a principal ideal domain.
+We express "free `R`-module of finite rank" as a module `M` which has a basis
+`b : ι → R`, where `ι` is a `fintype`.
+We call the cardinality of `ι` the rank of `M` in this file;
+it would be equal to `findim R M` if `R` is a field and `M` is a vector space.
+
+## Main results
+
+ - `submodule.induction_on_rank`: if `M` is free and finitely generated,
+   if `P` holds for `⊥ : submodule R M` and if `P N` follows from `P N'`
+   for all `N'` that are of lower rank, then `P` holds on all submodules
+
+ - `submodule.exists_is_basis`: if `M` is free and finitely generated
+   and `R` is a PID, then `N : submodule R M` is free and finitely generated.
+   This is the first part of the structure theorem for modules.
+
+## Tags
+
+free module, finitely generated module, rank, structure theorem
+
+-/
+
+open_locale big_operators
+
+section comm_ring
+
+universes u v
+
+variables {R : Type u} {M : Type v} [comm_ring R] [add_comm_group M] [module R M]
+variables {ι : Type*} {b : ι → M} (hb : is_basis R b)
+
+open submodule.is_principal
+
+lemma eq_bot_of_rank_eq_zero [no_zero_divisors R] (hb : is_basis R b) (N : submodule R M)
+  (rank_eq : ∀ {m : ℕ} (v : fin m → N),
```

Comment:
Does it make sense to have a similar lemma for `fintype`s instead of only `fin m`?

Reply:
It's a bit awkward because the condition would look like
```lean
  (rank_eq : ∀ {ι : Type*} {ft : fintype ι} (v : ι → N),
    linear_independent R (coe ∘ v : ι → M) → @fintype.card ι ft = 0) :
```
and the counterexample becomes `fintype.card (ulift (fin 1)) ≠ 0` instead of just `1 ≠ 0`. I'll add a version for `finset N`s, let me know if you still want a `fintype` version.

Reply:
(Also, using `fin` works well with the condition of `submodule.induction_on_rank_aux`. I recall trying to state that with `finset`s instead and it became a mess of `coe`s, `erase`s and `insert`s.)

Reply:
Bah, I can't figure out how to turn `sum_eq : ∑ (i : ↥↑{⟨x, hx⟩}), g i • ↑i = 0,` into `g ⟨⟨x, hx⟩, _⟩ = 0` and it's almost dinner time. Let me know if you still want a different variant, I'll take a fresh look at it tomorrow.

Reply:
Ooh well, we can always add it later. It's probably useful to have both variants.